src/IO/ReadBufferFromS3.h

```diff
@@ -33,8 +33,11 @@ class ReadBufferFromS3 : public SeekableReadBufferWithSize
     String key;
     UInt64 max_single_read_retries;
 
-    off_t offset = 0;
-    off_t read_until_position = 0;
+    /// These variables are atomic because they can be used for `logging only`
+    /// (where it is not important to get consistent result)
+    /// from separate thread other than the one which uses the buffer for s3 reading.
+    std::atomic<off_t> offset = 0;
```

Comment:
But why it cannot race with other fieds? In other words why the race can happens only with logging?

Reply:
> But why it cannot race with other fieds?

But why it should, if other fields are not accessed concurrently?
This is the only place, where buffer field is accessed concurrently, which I know is not good, but it is used only for adding info to exception message and it saved me a lot with debugging some errors, so I would rather not remove it.

Reply:
I'm looking at `nextImplStep`, and trying to understand why for the same `implementation_buffer` only `getRemainingReadRange()` can be called concurrently.

AFAICS one thread was downloading, while another, concurrently, was waiting for the download to be finished, and it should not call anything on this `implementation_buffer`, no?
I guess some exception can lead to this? Or what did I miss?

Reply:
The concurrency can be in the following case:

file is still being downloaded by someone, but we do not necessarily wait for the full download if we know that some part of it, which we need, is already downloaded:
https://github.com/ClickHouse/ClickHouse/blob/e3d8da26b77da09feabbda332fe7e6ad56fa203e/src/Disks/IO/CachedReadBufferFromRemoteFS.cpp#L198-L210

Then `getRemainingReadRange()` is only used in case of an exception from `nextImpl` and it's result is added to exception message. So here is a race on remaining read range, because someone can still be downloading. This race is ok because it is just for debug and better understanding of the current exception.

Reply:
Ok, thanks.

So, AFAIU it is only possible if `getImplementationBuffer` will fail, and so `implementation_buffer` will not be updated?

Reply:
> is only possible if getImplementationBuffer will fail

no, it can succeed, `implementation_buffer` will be updated. Then if any exception happens within `nextImplStep` it will be caught in `nextImpl` and a cache info message is appended to exception message.

Reply:
But if it will update `implementation_buffer`, then it will be an instance of `ReadBufferFromFile`, not `ReadBufferFromS3` there the race is.

Reply:
Actually not, because the case I described is not applied to `ReadBufferFromFile` as it is not shared between file segments like `ReadBufferFromS3`.